File: julia/src/Antybrowser.jl
module Antybrowser

using HTTP
using JSON3

export AntybrowserClient
export get_status, get_settings, get_sync_status, refresh_sync
export get_profiles, create_profile, update_profile, delete_profile, start_profile, stop_profile, duplicate_profile
export get_automations, run_automation
export get_groups, create_group, update_group, delete_group
export get_proxies, create_proxy, check_proxy, delete_proxy
export get_extensions, delete_extension, get_profile_extensions, set_profile_extensions

struct AntybrowserError <: Exception
    message::String
    status_code::Union{Int,Nothing}
    body::Union{String,Nothing}
end

AntybrowserError(msg::String) = AntybrowserError(msg, nothing, nothing)

Base.showerror(io::IO, e::AntybrowserError) = print(io, "AntybrowserError: ", e.message)

struct AntybrowserClient
    api_key::String
    base_url::String
end

function AntybrowserClient(api_key::String; port::Int=5173)
    AntybrowserClient(api_key, "http://127.0.0.1:$port")
end

function _request(c::AntybrowserClient, method::Symbol, path::String, body=nothing)
    url = c.base_url * path
    headers = Dict("x-api-key" => c.api_key, "Content-Type" => "application/json")

    resp = if method == :GET
        HTTP.get(url; headers=headers)
    elseif method == :POST
        HTTP.post(url; headers=headers, body=body !== nothing ? JSON3.write(body) : "{}")
    elseif method == :PUT
        HTTP.put(url; headers=headers, body=body !== nothing ? JSON3.write(body) : "{}")
    elseif method == :DELETE
        HTTP.delete(url; headers=headers)
    end

    if resp.status >= 400
        throw(AntybrowserError("HTTP $(resp.status): $(String(resp.body))", resp.status, String(resp.body)))
    end
    JSON3.read(String(resp.body))
end

# System
get_status(c::AntybrowserClient) = _request(c, :GET, "/api/status")
get_settings(c::AntybrowserClient) = _request(c, :GET, "/api/settings")
get_sync_status(c::AntybrowserClient) = _request(c, :GET, "/api/sync/status")
refresh_sync(c::AntybrowserClient, profile_id=nothing) = _request(c, :POST, "/api/sync/refresh", profile_id !== nothing ? Dict("profileId" => profile_id) : Dict())

# Profiles
get_profiles(c::AntybrowserClient) = _request(c, :GET, "/api/profiles")
create_profile(c::AntybrowserClient, data) = _request(c, :POST, "/api/profiles", data)
update_profile(c::AntybrowserClient, id, data) = _request(c, :PUT, "/api/profiles/$id", data)
delete_profile(c::AntybrowserClient, id) = _request(c, :DELETE, "/api/profiles/$id")
start_profile(c::AntybrowserClient, id) = _request(c, :POST, "/api/profiles/$id/start")
stop_profile(c::AntybrowserClient, id) = _request(c, :POST, "/api/profiles/$id/stop")
duplicate_profile(c::AntybrowserClient, id, name=nothing) = _request(c, :POST, "/api/profiles/$id/duplicate", name !== nothing ? Dict("name" => name) : Dict())

# Automations
get_automations(c::AntybrowserClient) = _request(c, :GET, "/api/automations")
run_automation(c::AntybrowserClient, id, profile_id) = _request(c, :POST, "/api/automations/$id/run", Dict("profileId" => profile_id))

# Groups
get_groups(c::AntybrowserClient) = _request(c, :GET, "/api/groups")
create_group(c::AntybrowserClient, data) = _request(c, :POST, "/api/groups", data)
update_group(c::AntybrowserClient, id, data) = _request(c, :PUT, "/api/groups/$id", data)
delete_group(c::AntybrowserClient, id) = _request(c, :DELETE, "/api/groups/$id")

# Proxies
get_proxies(c::AntybrowserClient) = _request(c, :GET, "/api/proxies")
create_proxy(c::AntybrowserClient, data) = _request(c, :POST, "/api/proxies", data)
check_proxy(c::AntybrowserClient, data) = _request(c, :POST, "/api/proxies/check", data)
delete_proxy(c::AntybrowserClient, id) = _request(c, :DELETE, "/api/proxies/$id")

# Extensions
get_extensions(c::AntybrowserClient) = _request(c, :GET, "/api/extensions")
delete_extension(c::AntybrowserClient, id) = _request(c, :DELETE, "/api/extensions/$id")
get_profile_extensions(c::AntybrowserClient, profile_id) = _request(c, :GET, "/api/profiles/$profile_id/extensions")
set_profile_extensions(c::AntybrowserClient, profile_id, extension_ids) = _request(c, :POST, "/api/profiles/$profile_id/extensions", Dict("extensionIds" => extension_ids))

end

File: r/R/antybrowser.R
#' Antybrowser R Client
#'
#' Official R client for the Antybrowser Local API.
#'
#' @docType package
#' @name antybrowser-package
NULL

#' Create an Antybrowser client
#'
#' @param api_key Your API key from Antybrowser settings
#' @param port API port (default 5173)
#' @return An AntybrowserClient object
#' @export
AntybrowserClient <- function(api_key, port = 5173) {
  structure(
    list(
      api_key = api_key,
      base_url = paste0("http://127.0.0.1:", port)
    ),
    class = "AntybrowserClient"
  )
}

.do_request <- function(client, method, path, body = NULL) {
  url <- paste0(client$base_url, path)
  headers <- c("x-api-key" = client$api_key, "Content-Type" = "application/json")

  response <- switch(method,
    GET = httr::GET(url, httr::add_headers(.headers = headers)),
    POST = httr::POST(url, httr::add_headers(.headers = headers), body = if (!is.null(body)) jsonlite::toJSON(body, auto_unbox = TRUE) else "{}", encode = "raw"),
    PUT = httr::PUT(url, httr::add_headers(.headers = headers), body = if (!is.null(body)) jsonlite::toJSON(body, auto_unbox = TRUE) else "{}", encode = "raw"),
    DELETE = httr::DELETE(url, httr::add_headers(.headers = headers))
  )

  if (httr::status_code(response) >= 400) {
    stop(paste("API error", httr::status_code(response), ":", httr::content(response, "text")))
  }
  jsonlite::fromJSON(httr::content(response, "text"), simplifyVector = FALSE)
}

#' @rdname AntybrowserClient
#' @export
get_status <- function(client) .do_request(client, "GET", "/api/status")
#' @export
get_settings <- function(client) .do_request(client, "GET", "/api/settings")
#' @export
get_sync_status <- function(client) .do_request(client, "GET", "/api/sync/status")
#' @export
refresh_sync <- function(client, profile_id = NULL) {
  body <- if (!is.null(profile_id)) list(profileId = profile_id) else list()
  .do_request(client, "POST", "/api/sync/refresh", body)
}

#' @rdname AntybrowserClient
#' @export
get_profiles <- function(client) .do_request(client, "GET", "/api/profiles")
#' @export
create_profile <- function(client, data) .do_request(client, "POST", "/api/profiles", data)
#' @export
update_profile <- function(client, id, data) .do_request(client, "PUT", paste0("/api/profiles/", id), data)
#' @export
delete_profile <- function(client, id) .do_request(client, "DELETE", paste0("/api/profiles/", id))
#' @export
start_profile <- function(client, id) .do_request(client, "POST", paste0("/api/profiles/", id, "/start"))
#' @export
stop_profile <- function(client, id) .do_request(client, "POST", paste0("/api/profiles/", id, "/stop"))
#' @export
duplicate_profile <- function(client, id, name = NULL) {
  body <- if (!is.null(name)) list(name = name) else list()
  .do_request(client, "POST", paste0("/api/profiles/", id, "/duplicate"), body)
}

#' @rdname AntybrowserClient
#' @export
get_automations <- function(client) .do_request(client, "GET", "/api/automations")
#' @export
run_automation <- function(client, id, profile_id) .do_request(client, "POST", paste0("/api/automations/", id, "/run"), list(profileId = profile_id))

#' @rdname AntybrowserClient
#' @export
get_groups <- function(client) .do_request(client, "GET", "/api/groups")
#' @export
create_group <- function(client, data) .do_request(client, "POST", "/api/groups", data)
#' @export
update_group <- function(client, id, data) .do_request(client, "PUT", paste0("/api/groups/", id), data)
#' @export
delete_group <- function(client, id) .do_request(client, "DELETE", paste0("/api/groups/", id))

#' @rdname AntybrowserClient
#' @export
get_proxies <- function(client) .do_request(client, "GET", "/api/proxies")
#' @export
create_proxy <- function(client, data) .do_request(client, "POST", "/api/proxies", data)
#' @export
check_proxy <- function(client, data) .do_request(client, "POST", "/api/proxies/check", data)
#' @export
delete_proxy <- function(client, id) .do_request(client, "DELETE", paste0("/api/proxies/", id))

#' @rdname AntybrowserClient
#' @export
get_extensions <- function(client) .do_request(client, "GET", "/api/extensions")
#' @export
delete_extension <- function(client, id) .do_request(client, "DELETE", paste0("/api/extensions/", id))
#' @export
get_profile_extensions <- function(client, profile_id) .do_request(client, "GET", paste0("/api/profiles/", profile_id, "/extensions"))
#' @export
set_profile_extensions <- function(client, profile_id, extension_ids) {
  .do_request(client, "POST", paste0("/api/profiles/", profile_id, "/extensions"), list(extensionIds = extension_ids))
}
